doc/man3/ASN1_item_sign.pod

```diff
@@ -0,0 +1,161 @@
+
+ASN1_item_verify_ex() is used to verify the signature I<signature> of internal
+data I<data> using the public key I<pkey> and algorithm identifier I<alg>.
+The data that is verified is formed by taking the data object in I<data> and
+converting it to der format using the ASN.1 structure I<it>.
+The B<OSSL_LIB_CTX> specified in I<libctx> and the property query string
+specified in I<props> are used when searching for algorithms in providers.
+The optional parameter I<id> can be NULL, but can be set for special key types.
+See EVP_PKEY_CTX_set1_id() for further info.
+
+ASN1_item_verify() is similar to ASN1_item_verify_ex() but uses default values of
+NULL for the I<id>, I<libctx> and I<propq>.
+
+ASN1_item_verify_ctx() is similiar to ASN1_item_verify() but uses the parameters
+contained in digest context I<ctx>.
+
+
+=head1 RETURN VALUES
+
+All sign functions return the size of the signature in bytes for success and
+zero for failure.
+
+All verify functions return 1 if the signature is valid and 0 if the signature
+check fails. If the signature could not be checked at all because it was
+ill-formed or some other error occurred then -1 is returned.
+
+=head1 EXAMPLES
+
+In the following example an 'MyObject' object is signed using the key contained
+in an md_ctx. The signature is written to MyObject.signature.
+
+ #include <openssl/x509.h>
+
+ /* An object used to store the ASN1 data fields to sign */
+ typedef struct MySignObject_st
+ {
+     ASN1_INTEGER version;
+     X509_ALGOR sig_alg;
+ } MySignObject;
+
+ /*
+  * A higher level object containing the ASN1 fields, signature alg and
+  * output signature.
+  */
+ typedef struct MyObject_st
+ {
+     MySignObject info;
+     X509_ALGOR sig_alg;
+     ASN1_BIT_STRING signature;
+ } MyObject;
+
+ ASN1_SEQUENCE_cb(MySignObject, NULL) = {
+     ASN1_EMBED(MySignObject, version, ASN1_INTEGER),
+     ASN1_EMBED(MySignObject, sig_alg, X509_ALGOR),
+ } static_ASN1_SEQUENCE_END_cb(MySignObject, MySignObject)
+
+ static int test_asn1_item_sign(MyObject *obj, EVP_PKEY *pkey, EVP_MD *md)
+ {
+     EVP_PKEY *pkey = NULL;
+     EVP_PKEY_CTX *gctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL);
+
+     /* 'it' contains the mapping between ASN.1 data and an object MySignObject */
+     const ASN1_ITEM *it = ASN1_ITEM_rptr(MySignObject);
+     EVP_MD_CTX *mdctx = EVP_MD_CTX_new();
+
+     memset(obj, 0, sizeof(*obj));
+     EVP_DigestSignInit(mdctx, NULL, md, NULL, pkey);
+     /*
+      * obj->info is the 'MySignObject' object that will be
+      * converted into DER data and then signed.
+      * obj->signature will contain the output signature.
+      * obj->sig_alg is filled with the private key's signing algorithm id.
+      * obj->info.sig_alg is another copy of the signing algorithm id that sits
+      * within the MyObject object.
+      */
+     return ASN1_item_sign_ctx(it, &obj->sig_alg, &obj->info.sig_alg,
+                               &obj->signature, &obj->info, mdctx);
+ }
```

Comment:
The memset() isn't making much sense to me, you're removing the data from obj->info.

I wonder if it's useful to also created an ASN1 object for MyObject and then did i2d on that to convert the whole thing to ASN1.

Reply:
The example should probably also cover verify

Reply:
Thanks for the input.. That was a useful suggestion.
See what you think of the example now.